{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybiomart import Dataset\n",
    "import pandas as pd\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\joaop\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\pybiomart\\dataset.py:269: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  result = pd.read_csv(StringIO(response.text), sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "dataset = Dataset(name='hsapiens_gene_ensembl', host='http://www.ensembl.org')\n",
    "\n",
    "\n",
    "df = dataset.query(attributes=['ensembl_gene_id', 'external_gene_name', 'ensembl_transcript_id', \n",
    "                               'transcript_is_canonical','chromosome_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Ensembl Canonical</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277076</th>\n",
       "      <td>ENSG00000288678</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000679970</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277077</th>\n",
       "      <td>ENSG00000288678</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000710329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277078</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277079</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277080</th>\n",
       "      <td>ENSG00000290826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000642116</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>277081 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID Gene name Transcript stable ID  Ensembl Canonical  \\\n",
       "0       ENSG00000210049     MT-TF      ENST00000387314                1.0   \n",
       "1       ENSG00000211459   MT-RNR1      ENST00000389680                1.0   \n",
       "2       ENSG00000210077     MT-TV      ENST00000387342                1.0   \n",
       "3       ENSG00000210082   MT-RNR2      ENST00000387347                1.0   \n",
       "4       ENSG00000209082    MT-TL1      ENST00000386347                1.0   \n",
       "...                 ...       ...                  ...                ...   \n",
       "277076  ENSG00000288678       NaN      ENST00000679970                1.0   \n",
       "277077  ENSG00000288678       NaN      ENST00000710329                NaN   \n",
       "277078  ENSG00000290825   DDX11L2      ENST00000456328                1.0   \n",
       "277079  ENSG00000227232    WASH7P      ENST00000488147                1.0   \n",
       "277080  ENSG00000290826       NaN      ENST00000642116                1.0   \n",
       "\n",
       "       Chromosome/scaffold name  \n",
       "0                            MT  \n",
       "1                            MT  \n",
       "2                            MT  \n",
       "3                            MT  \n",
       "4                            MT  \n",
       "...                         ...  \n",
       "277076                        1  \n",
       "277077                        1  \n",
       "277078                        1  \n",
       "277079                        1  \n",
       "277080                        1  \n",
       "\n",
       "[277081 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MT', 'GL000009.2', 'GL000220.1', 'KI270731.1', 'GL000195.1',\n",
       "       'KI270442.1', 'GL000205.2', 'KI270744.1', 'KI270750.1',\n",
       "       'GL000219.1', 'KI270734.1', 'GL000194.1', 'GL000213.1',\n",
       "       'KI270711.1', 'KI270721.1', 'KI270713.1', 'KI270726.1',\n",
       "       'KI270728.1', 'GL000218.1', 'KI270733.1', 'HG1309_PATCH',\n",
       "       'GL000225.1', 'GL000216.2', 'HG1395_PATCH', 'HG1485_PATCH',\n",
       "       'HG151_NOVEL_TEST', 'Y', 'HG1524_PATCH', 'HG152_PATCH',\n",
       "       'HG1532_PATCH', 'HG1535_PATCH', 'HG1651_PATCH', 'HG1708_PATCH',\n",
       "       'HG1815_PATCH', 'HG1832_PATCH', 'HG1_PATCH', 'HG2021_PATCH',\n",
       "       'HG2002_PATCH', 'HG2030_PATCH', 'HG2022_PATCH', 'HG2047_PATCH',\n",
       "       'HG2023_PATCH', 'HG2052_PATCH', 'HG2031_PATCH', 'HG2058_PATCH',\n",
       "       'HG2046_PATCH', 'HG2062_PATCH', 'HG2057_PATCH', 'HG2066_PATCH',\n",
       "       'HG2060_PATCH', '21', 'HG2063_PATCH', 'HG2067_PATCH',\n",
       "       'HG2072_PATCH', '13', 'HG2069_PATCH', 'HG2088_PATCH',\n",
       "       'HG2095_PATCH', 'HG2077_PATCH', 'HG2111_PATCH', 'HG2087_PATCH',\n",
       "       'HG2115_PATCH', 'HG2104_PATCH', 'HG2118_PATCH', 'HG2114_PATCH',\n",
       "       'HG2128_PATCH', 'HG2116_PATCH', 'HG2140_PATCH', 'HG2121_PATCH',\n",
       "       'HG2158_PATCH', 'HG2133_PATCH', 'HG2191_PATCH', 'HG2176_PATCH',\n",
       "       '18', 'HG2198_PATCH', 'HG2219_PATCH', 'HG2213_PATCH',\n",
       "       'HG2231_HG2496_PATCH', 'HG2233_PATCH', 'HG2217_PATCH',\n",
       "       'HG2235_PATCH', 'HG2225_PATCH', 'HG2232_PATCH', 'HG2236_PATCH',\n",
       "       'HG2246_HG2248_HG2276_PATCH', 'HG2239_PATCH', 'HG2249_PATCH',\n",
       "       'HG2263_PATCH', 'HG2247_PATCH', 'HG2265_PATCH', 'HG2251_PATCH',\n",
       "       'HG2267_PATCH', 'HG2264_PATCH', 'HG2280_PATCH', 'HG2266_PATCH',\n",
       "       '22', 'HG2275_PATCH', 'HG2290_PATCH', 'HG2291_PATCH',\n",
       "       'HG2285_HG106_HG2252_PATCH', 'HG2334_PATCH', 'HG2288_HG2289_PATCH',\n",
       "       'HG2408_PATCH', 'HG2308_PATCH', 'HG2412_PATCH', 'HG2442_PATCH',\n",
       "       'HG2365_PATCH', 'HG2461_PATCH', '20', 'HG2407_PATCH',\n",
       "       'HG2476_PATCH', 'HG2499_PATCH', 'HG2419_PATCH', 'HG2509_PATCH',\n",
       "       'HG2510_PATCH', 'HG2471_PATCH', 'HG2512_PATCH', 'HG2494_PATCH',\n",
       "       'HG2515_PATCH', 'HG2511_PATCH', 'HG2525_PATCH', 'HG2513_PATCH',\n",
       "       'HG2527_PATCH', 'HG2521_PATCH', 'HG2554_PATCH',\n",
       "       'HG2526_HG2573_PATCH', 'HG2569_PATCH', 'HG2571_PATCH',\n",
       "       'HG2541_PATCH', 'HG2577_PATCH', 'HG2568_PATCH', 'HG2580_PATCH',\n",
       "       'HG2576_PATCH', 'HG287_PATCH', 'HG2578_PATCH', 'HG30_PATCH',\n",
       "       'HG26_PATCH', 'X', 'HG405_PATCH', 'HG28_PATCH', '14', '15',\n",
       "       'HG439_PATCH', 'HG401_PATCH', 'HG545_PATCH', '10', 'HG410_PATCH',\n",
       "       'HG705_PATCH', 'HG721_PATCH', 'HG699_PATCH', 'HG926_PATCH', '9',\n",
       "       'HG708_PATCH', '8', 'HSCHR10_1_CTG1', 'HSCHR10_1_CTG3',\n",
       "       'HG76_PATCH', 'HSCHR10_1_CTG4', 'HSCHR11_1_CTG1_2', 'HG986_PATCH',\n",
       "       'HSCHR11_1_CTG5', 'HSCHR10_1_CTG2', 'HSCHR11_1_CTG7',\n",
       "       'HSCHR11_1_CTG3_1', 'HSCHR11_1_CTG6', 'HSCHR11_2_CTG1',\n",
       "       'HSCHR11_2_CTG8', 'HSCHR11_1_CTG8', 'HSCHR11_3_CTG1',\n",
       "       'HSCHR11_2_CTG1_1', '16', 'HSCHR11_2_CTG3_1', '4',\n",
       "       'HSCHR12_1_CTG1', 'HSCHR12_2_CTG2_1', '5', 'HSCHR12_3_CTG2',\n",
       "       'HSCHR12_1_CTG2_1', 'HSCHR12_2_CTG2', 'HSCHR12_4_CTG2_1',\n",
       "       'HSCHR12_5_CTG2', 'HSCHR12_3_CTG2_1', '7', 'HSCHR12_6_CTG2_1',\n",
       "       'HSCHR12_4_CTG2', 'HSCHR12_5_CTG2_1', 'HSCHR13_1_CTG1',\n",
       "       'HSCHR13_1_CTG3', 'HSCHR12_9_CTG2_1', 'HSCHR13_1_CTG5', '6',\n",
       "       'HSCHR14_1_CTG1', 'HSCHR14_7_CTG1', 'HSCHR14_2_CTG1',\n",
       "       'HSCHR14_3_CTG1', 'HSCHR15_1_CTG1', 'HSCHR15_1_CTG3',\n",
       "       'HSCHR14_8_CTG1', 'HSCHR15_2_CTG8', 'HSCHR15_1_CTG8',\n",
       "       'HSCHR15_3_CTG8', 'HSCHR15_2_CTG3', 'HSCHR15_5_CTG8', '19',\n",
       "       'HSCHR15_3_CTG3', 'HSCHR15_9_CTG8', 'HSCHR15_4_CTG8',\n",
       "       'HSCHR16_1_CTG3_1', '12', 'HSCHR16_2_CTG3_1', 'HSCHR15_6_CTG8',\n",
       "       'HSCHR16_4_CTG3_1', 'HSCHR16_1_CTG1', 'HSCHR16_5_CTG1',\n",
       "       'HSCHR16_5_CTG3_1', 'HSCHR16_CTG2', 'HSCHR16_3_CTG1',\n",
       "       'HSCHR17_12_CTG4', 'HSCHR16_4_CTG1', 'HSCHR17_1_CTG1',\n",
       "       'HSCHR17_10_CTG4', 'HSCHR17_1_CTG4', 'HSCHR17_13_CTG4',\n",
       "       'HSCHR17_1_CTG9', 'HSCHR17_1_CTG2', 'HSCHR17_2_CTG2', '11', '3',\n",
       "       'HSCHR17_1_CTG5', 'HSCHR17_2_CTG5', '17', 'HSCHR17_2_CTG1',\n",
       "       'HSCHR17_3_CTG2', 'HSCHR17_2_CTG4', 'HSCHR17_4_CTG4',\n",
       "       'HSCHR17_3_CTG1', 'HSCHR17_6_CTG4', 'HSCHR17_3_CTG4',\n",
       "       'HSCHR17_7_CTG4', 'HSCHR17_5_CTG4', 'HSCHR17_8_CTG4',\n",
       "       'HSCHR17_9_CTG4', 'HSCHR18_1_CTG1', 'HSCHR18_1_CTG1_1',\n",
       "       'HSCHR18_2_CTG1_1', 'HSCHR18_2_CTG2', 'HSCHR18_1_CTG2_1',\n",
       "       'HSCHR18_5_CTG1_1', 'HSCHR18_2_CTG2_1', 'HSCHR18_ALT2_CTG2_1',\n",
       "       'HSCHR18_3_CTG2_1', 'HSCHR18_ALT21_CTG2_1',\n",
       "       'HSCHR19KIR_0019-4656-A_CTG3_1', 'HSCHR19KIR_0010-5217-AB_CTG3_1',\n",
       "       'HSCHR19KIR_502960008-2_CTG3_1', 'HSCHR19KIR_0019-4656-B_CTG3_1',\n",
       "       'HSCHR19KIR_7191059-2_CTG3_1', 'HSCHR19KIR_502960008-1_CTG3_1',\n",
       "       'HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1',\n",
       "       'HSCHR19KIR_7191059-1_CTG3_1', 'HSCHR19KIR_CA01-TA01_1_CTG3_1',\n",
       "       'HSCHR19KIR_ABC08_A1_HAP_CTG3_1', 'HSCHR19KIR_CA01-TB01_CTG3_1',\n",
       "       'HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1', 'HSCHR19KIR_CA04_CTG3_1',\n",
       "       'HSCHR19KIR_CA01-TA01_2_CTG3_1', 'HSCHR19KIR_FH05_A_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_CA01-TB04_CTG3_1', 'HSCHR19KIR_FH06_BA1_HAP_CTG3_1',\n",
       "       '2', 'HSCHR19KIR_FH08_A_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_FH05_B_HAP_CTG3_1', 'HSCHR19KIR_FH13_A_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_FH06_A_HAP_CTG3_1', 'HSCHR19KIR_FH15_A_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_FH08_BAX_HAP_CTG3_1', 'HSCHR19KIR_G085_A_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_FH13_BA2_HAP_CTG3_1', 'HSCHR19KIR_G248_A_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_FH15_B_HAP_CTG3_1', 'HSCHR19KIR_GRC212_AB_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_G085_BA1_HAP_CTG3_1', 'HSCHR19KIR_HG2393_CTG3_1',\n",
       "       'HSCHR19KIR_G248_BA2_HAP_CTG3_1', 'HSCHR19KIR_HG2396_CTG3_1',\n",
       "       'HSCHR19KIR_GRC212_BA1_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1', 'HSCHR19KIR_HG2394_CTG3_1',\n",
       "       'HSCHR19KIR_RSH_A_HAP_CTG3_1', 'HSCHR19KIR_LUCE_A_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_T7526_A_HAP_CTG3_1', 'HSCHR19KIR_RP5_B_HAP_CTG3_1',\n",
       "       'HSCHR19KIR_RSH_BA2_HAP_CTG3_1', '1', 'HSCHR19LRC_COX1_CTG3_1',\n",
       "       'HSCHR19KIR_T7526_BDEL_HAP_CTG3_1', 'HSCHR19LRC_COX2_CTG3_1',\n",
       "       'HSCHR19LRC_LRC_S_CTG3_1', 'HSCHR19LRC_LRC_T_CTG3_1',\n",
       "       'HSCHR19LRC_LRC_I_CTG3_1', 'HSCHR19LRC_LRC_J_CTG3_1',\n",
       "       'HSCHR19LRC_PGF2_CTG3_1', 'HSCHR19LRC_PGF1_CTG3_1',\n",
       "       'HSCHR19_2_CTG2', 'HSCHR19_2_CTG3_1', 'KI270727.1',\n",
       "       'HSCHR19_3_CTG3_1', 'HSCHR19_1_CTG3_1', 'HG1012_PATCH',\n",
       "       'HSCHR19_4_CTG2', 'HSCHR19_4_CTG3_1', 'HSCHR19_5_CTG2',\n",
       "       'HG1046_PATCH', 'HSCHR1_12_CTG3', 'HG1047_PATCH', 'HSCHR1_1_CTG3',\n",
       "       'HG107_HG2565_PATCH', 'HSCHR1_2_CTG31', 'HG109_PATCH',\n",
       "       'HSCHR1_2_CTG3', 'HSCHR1_3_CTG31', 'HSCHR1_2_CTG32_1',\n",
       "       'HSCHR1_4_CTG3', 'HG1206_PATCH', 'HSCHR1_3_CTG32_1', 'HG126_PATCH',\n",
       "       'HSCHR1_5_CTG31', 'HG1277_PATCH', 'HSCHR1_6_CTG31',\n",
       "       'HSCHR1_5_CTG32_1', 'HSCHR1_8_CTG3', 'HSCHR20_1_CTG1',\n",
       "       'HG1298_PATCH', 'HSCHR20_1_CTG3', 'HSCHR1_ALT2_1_CTG32_1',\n",
       "       'HSCHR20_1_CTG4', 'HSCHR21_3_CTG1_1', 'HG1311_HG2539_PATCH',\n",
       "       'HSCHR21_2_CTG1_1', 'HSCHR21_6_CTG1_1', 'HG1320_PATCH',\n",
       "       'HSCHR21_5_CTG2', 'HSCHR22_1_CTG2', 'HG1342_HG2282_PATCH',\n",
       "       'HSCHR22_1_CTG4', 'HSCHR22_1_CTG1', 'HG1343_HG173_HG459_PATCH',\n",
       "       'HSCHR22_1_CTG5', 'HSCHR22_1_CTG7', 'HSCHR22_1_CTG6',\n",
       "       'HG1362_PATCH', 'HSCHR22_5_CTG1', 'HSCHR22_3_CTG1', 'HG1369_PATCH',\n",
       "       'HSCHR22_8_CTG1', 'HSCHR22_6_CTG1', 'HSCHR2_11_CTG7_2',\n",
       "       'HG1384_PATCH', 'HSCHR2_10_CTG7_2', 'HG1398_PATCH',\n",
       "       'HSCHR2_1_CTG5', 'HSCHR2_1_CTG1', 'HSCHR2_1_CTG7_2',\n",
       "       'HSCHR2_1_CTG15', 'HG142_HG150_NOVEL_TEST', 'HSCHR2_2_CTG7',\n",
       "       'HSCHR2_2_CTG1', 'HG1445_PATCH', 'HSCHR2_3_CTG15',\n",
       "       'HSCHR2_2_CTG7_2', 'HSCHR19_1_CTG2', 'HSCHR2_3_CTG7_2',\n",
       "       'HSCHR19_3_CTG2', 'HSCHR2_7_CTG7_2', 'HSCHR19_6_CTG2',\n",
       "       'HSCHR3_1_CTG2_1', 'HSCHR2_6_CTG7_2', 'HSCHR2_1_CTG7',\n",
       "       'HSCHR1_1_CTG11', 'HSCHR2_3_CTG1', 'HSCHR3_2_CTG3',\n",
       "       'HSCHR3_2_CTG2_1', 'HSCHR1_1_CTG31', 'HSCHR2_6_CTG1',\n",
       "       'HSCHR3_4_CTG2_1', 'HSCHR1_1_CTG32_1', 'HSCHR3_3_CTG3',\n",
       "       'HSCHR3_1_CTG1', 'HSCHR3_5_CTG3', 'HSCHR1_3_CTG3', 'HSCHR3_3_CTG1',\n",
       "       'HSCHR1_4_CTG31', 'HSCHR3_4_CTG1', 'HSCHR3_9_CTG2_1',\n",
       "       'HSCHR1_5_CTG3', 'HSCHR3_5_CTG1', 'HSCHR3_8_CTG3',\n",
       "       'HSCHR4_12_CTG12', 'HSCHR1_6_CTG3', 'HSCHR3_6_CTG2_1',\n",
       "       'HSCHR4_1_CTG6', 'HSCHR1_9_CTG3', 'HSCHR4_1_CTG4',\n",
       "       'HSCHR20_1_CTG2', 'HSCHR3_7_CTG3', 'HSCHR4_2_CTG12',\n",
       "       'HSCHR4_4_CTG12', 'HSCHR4_3_CTG12', 'HSCHR4_7_CTG12',\n",
       "       'HSCHR21_4_CTG1_1', 'HSCHR4_1_CTG12', 'HSCHR4_8_CTG12',\n",
       "       'HSCHR5_1_CTG1', 'HSCHR21_8_CTG1_1', 'HSCHR4_1_CTG9',\n",
       "       'HSCHR5_1_CTG1_1', 'HSCHR5_2_CTG1_1', 'HSCHR22_1_CTG3',\n",
       "       'HSCHR4_5_CTG12', 'HSCHR5_3_CTG1', 'HSCHR4_6_CTG12',\n",
       "       'HSCHR22_2_CTG1', 'HSCHR5_4_CTG1_1', 'HSCHR5_10_CTG1',\n",
       "       'HSCHR22_4_CTG1', 'HSCHR5_5_CTG1', 'HSCHR5_8_CTG1',\n",
       "       'HSCHR5_2_CTG1', 'HSCHR5_7_CTG1', 'HSCHR22_7_CTG1',\n",
       "       'HSCHR6_1_CTG2', 'HSCHR6_1_CTG1', 'HSCHR2_12_CTG7_2',\n",
       "       'HSCHR6_1_CTG6', 'HSCHR5_4_CTG1', 'HSCHR6_1_CTG5', 'HSCHR6_1_CTG9',\n",
       "       'HSCHR2_2_CTG15', 'HSCHR6_8_CTG1', 'HSCHR2_4_CTG1',\n",
       "       'HSCHR6_1_CTG3', 'HSCHR2_8_CTG7_2', 'HSCHR6_1_CTG7',\n",
       "       'HSCHR6_1_CTG8', 'HSCHR3_1_CTG3', 'HSCHR6_MHC_MANN_CTG1',\n",
       "       'HSCHR6_MHC_DBB_CTG1', 'HSCHR6_MHC_COX_CTG1', 'HSCHR3_4_CTG3',\n",
       "       'HSCHR3_5_CTG2_1', 'HSCHR3_6_CTG3', 'HSCHR3_8_CTG2_1',\n",
       "       'HSCHR3_9_CTG3', 'HSCHR4_11_CTG12', 'HSCHR7_3_CTG4_4',\n",
       "       'HSCHR7_4_CTG1', 'HSCHR8_1_CTG7', 'HSCHR4_2_CTG8_1',\n",
       "       'HSCHR8_1_CTG6', 'HSCHR8_3_CTG1', 'HSCHR8_2_CTG1',\n",
       "       'HSCHR4_9_CTG12', 'HSCHR8_4_CTG7', 'HSCHR5_1_CTG5',\n",
       "       'HSCHR8_3_CTG7', 'HSCHR7_3_CTG1', 'HSCHR5_2_CTG5', 'HSCHR8_1_CTG1',\n",
       "       'HSCHR8_8_CTG1', 'HSCHR5_3_CTG5', 'HSCHR8_4_CTG1', 'HSCHR8_9_CTG1',\n",
       "       'HSCHR9_1_CTG3', 'HSCHR8_5_CTG1', 'HSCHR5_6_CTG1', 'HSCHR9_1_CTG6',\n",
       "       'HSCHR9_1_CTG4', 'HSCHR8_6_CTG1', 'HSCHRX_2_CTG12',\n",
       "       'HSCHRX_1_CTG14', 'HSCHR8_7_CTG7', 'HSCHR6_1_CTG4',\n",
       "       'HSCHR9_1_CTG2', 'HSCHRX_2_CTG14', 'HSCHR6_MHC_APD_CTG1',\n",
       "       'HSCHRX_2_CTG3', 'HSCHR6_MHC_MCF_CTG1', 'HSCHR6_MHC_QBL_CTG1',\n",
       "       'HSCHRX_3_CTG3', 'HSCHR7_1_CTG1', 'HSCHR7_1_CTG4_4',\n",
       "       'HSCHR6_MHC_SSTO_CTG1', 'HSCHR7_1_CTG6', 'HSCHR7_1_CTG7',\n",
       "       'HSCHR7_2_CTG1', 'HSCHR7_2_CTG4_4', 'HSCHR7_2_CTG7',\n",
       "       'HSCHR7_2_CTG6', 'HSCHR7_3_CTG6', 'HSCHR8_2_CTG7', 'HSCHR8_5_CTG7',\n",
       "       'HSCHR8_7_CTG1', 'HSCHR9_1_CTG1', 'HSCHR9_1_CTG5', 'HSCHRX_1_CTG3',\n",
       "       2, 3, 1, 11, 17], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Chromosome/scaffold name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosomes = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'\n",
    "               , '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y', 'XY', 'MT']\n",
    "\n",
    "df = df[df['Chromosome/scaffold name'].isin(chromosomes)]\n",
    "df=df[df[\"Ensembl Canonical\"]==1].copy()\n",
    "df.to_csv(\"pybiomart_q.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Ensembl Canonical</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107187</th>\n",
       "      <td>ENSG00000184634</td>\n",
       "      <td>MED12</td>\n",
       "      <td>ENST00000374080</td>\n",
       "      <td>1.0</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID Gene name Transcript stable ID  Ensembl Canonical  \\\n",
       "107187  ENSG00000184634     MED12      ENST00000374080                1.0   \n",
       "\n",
       "       Chromosome/scaffold name  \n",
       "107187                        X  "
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test=df[df['Gene name']=='MED12']\n",
    "test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# concat Cr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [],
   "source": [
    "results=pd.read_csv('SVC.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>Probability_Class_0</th>\n",
       "      <th>Probability_Class_1</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UBR5</td>\n",
       "      <td>ENSG00000104517</td>\n",
       "      <td>0.025221</td>\n",
       "      <td>0.974779</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHD9</td>\n",
       "      <td>ENSG00000177200</td>\n",
       "      <td>0.029630</td>\n",
       "      <td>0.970370</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHD6</td>\n",
       "      <td>ENSG00000124177</td>\n",
       "      <td>0.042962</td>\n",
       "      <td>0.957038</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>ENSG00000171634</td>\n",
       "      <td>0.043160</td>\n",
       "      <td>0.956840</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCOR2</td>\n",
       "      <td>ENSG00000196498</td>\n",
       "      <td>0.043535</td>\n",
       "      <td>0.956465</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18471</th>\n",
       "      <td>ADAM20</td>\n",
       "      <td>ENSG00000134007</td>\n",
       "      <td>0.992088</td>\n",
       "      <td>0.007912</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18472</th>\n",
       "      <td>MSMP</td>\n",
       "      <td>ENSG00000215183</td>\n",
       "      <td>0.992300</td>\n",
       "      <td>0.007700</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18473</th>\n",
       "      <td>ADAM30</td>\n",
       "      <td>ENSG00000134249</td>\n",
       "      <td>0.992557</td>\n",
       "      <td>0.007443</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18474</th>\n",
       "      <td>ADAM21</td>\n",
       "      <td>ENSG00000139985</td>\n",
       "      <td>0.992964</td>\n",
       "      <td>0.007036</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18475</th>\n",
       "      <td>GBP5</td>\n",
       "      <td>ENSG00000154451</td>\n",
       "      <td>0.993178</td>\n",
       "      <td>0.006822</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18476 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                1  Probability_Class_0  Probability_Class_1  \\\n",
       "0        UBR5  ENSG00000104517             0.025221             0.974779   \n",
       "1        CHD9  ENSG00000177200             0.029630             0.970370   \n",
       "2        CHD6  ENSG00000124177             0.042962             0.957038   \n",
       "3        BPTF  ENSG00000171634             0.043160             0.956840   \n",
       "4       NCOR2  ENSG00000196498             0.043535             0.956465   \n",
       "...       ...              ...                  ...                  ...   \n",
       "18471  ADAM20  ENSG00000134007             0.992088             0.007912   \n",
       "18472    MSMP  ENSG00000215183             0.992300             0.007700   \n",
       "18473  ADAM30  ENSG00000134249             0.992557             0.007443   \n",
       "18474  ADAM21  ENSG00000139985             0.992964             0.007036   \n",
       "18475    GBP5  ENSG00000154451             0.993178             0.006822   \n",
       "\n",
       "      Chromosome/scaffold name  \n",
       "0                            8  \n",
       "1                           16  \n",
       "2                           20  \n",
       "3                           17  \n",
       "4                           12  \n",
       "...                        ...  \n",
       "18471                       14  \n",
       "18472                        9  \n",
       "18473                        1  \n",
       "18474                       14  \n",
       "18475                        1  \n",
       "\n",
       "[18476 rows x 5 columns]"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#concat df and results no index\n",
    "results=pd.merge(results,df, left_on='1', right_on='Gene stable ID')\n",
    "results=results[['0','1','Probability_Class_0','Probability_Class_1','Chromosome/scaffold name']]\n",
    "results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Concat SFARI\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>status</th>\n",
       "      <th>gene-symbol</th>\n",
       "      <th>gene-name</th>\n",
       "      <th>ensembl-id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genetic-category</th>\n",
       "      <th>gene-score</th>\n",
       "      <th>syndromic</th>\n",
       "      <th>eagle</th>\n",
       "      <th>number-of-reports</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>ABAT</td>\n",
       "      <td>4-aminobutyrate aminotransferase</td>\n",
       "      <td>ENSG00000183044</td>\n",
       "      <td>16</td>\n",
       "      <td>Rare Single Gene Mutation, Genetic Association</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>ABCA10</td>\n",
       "      <td>ATP-binding cassette, sub-family A (ABC1), mem...</td>\n",
       "      <td>ENSG00000154263</td>\n",
       "      <td>17</td>\n",
       "      <td>Rare Single Gene Mutation</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>ABCA13</td>\n",
       "      <td>ATP binding cassette subfamily A member 13</td>\n",
       "      <td>ENSG00000179869</td>\n",
       "      <td>7</td>\n",
       "      <td>Rare Single Gene Mutation, Functional</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ATP-binding cassette, sub-family A (ABC1), mem...</td>\n",
       "      <td>ENSG00000064687</td>\n",
       "      <td>19</td>\n",
       "      <td>Rare Single Gene Mutation</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>ABCE1</td>\n",
       "      <td>ATP binding cassette subfamily E member 1</td>\n",
       "      <td>ENSG00000164163</td>\n",
       "      <td>4</td>\n",
       "      <td>Rare Single Gene Mutation</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1157</th>\n",
       "      <td>9</td>\n",
       "      <td>ZNF774</td>\n",
       "      <td>Zinc finger protein 774</td>\n",
       "      <td>ENSG00000196391</td>\n",
       "      <td>15</td>\n",
       "      <td>Rare Single Gene Mutation</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1158</th>\n",
       "      <td>9</td>\n",
       "      <td>ZNF804A</td>\n",
       "      <td>Zinc finger protein 804A</td>\n",
       "      <td>ENSG00000170396</td>\n",
       "      <td>2</td>\n",
       "      <td>Rare Single Gene Mutation, Genetic Association...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1159</th>\n",
       "      <td>9</td>\n",
       "      <td>ZNF827</td>\n",
       "      <td>Zinc finger protein 827</td>\n",
       "      <td>ENSG00000151612</td>\n",
       "      <td>4</td>\n",
       "      <td>Rare Single Gene Mutation, Genetic Association</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>9</td>\n",
       "      <td>ZSWIM6</td>\n",
       "      <td>zinc finger SWIM-type containing 6</td>\n",
       "      <td>ENSG00000130449</td>\n",
       "      <td>5</td>\n",
       "      <td>Rare Single Gene Mutation, Syndromic, Genetic ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>9</td>\n",
       "      <td>ZWILCH</td>\n",
       "      <td>zwilchkinetochore protein</td>\n",
       "      <td>ENSG00000174442</td>\n",
       "      <td>15</td>\n",
       "      <td>Rare Single Gene Mutation</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1162 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      status gene-symbol                                          gene-name  \\\n",
       "0          9        ABAT                   4-aminobutyrate aminotransferase   \n",
       "1          9      ABCA10  ATP-binding cassette, sub-family A (ABC1), mem...   \n",
       "2          9      ABCA13         ATP binding cassette subfamily A member 13   \n",
       "3          9       ABCA7  ATP-binding cassette, sub-family A (ABC1), mem...   \n",
       "4          9       ABCE1          ATP binding cassette subfamily E member 1   \n",
       "...      ...         ...                                                ...   \n",
       "1157       9      ZNF774                            Zinc finger protein 774   \n",
       "1158       9     ZNF804A                           Zinc finger protein 804A   \n",
       "1159       9      ZNF827                            Zinc finger protein 827   \n",
       "1160       9      ZSWIM6                 zinc finger SWIM-type containing 6   \n",
       "1161       9      ZWILCH                          zwilchkinetochore protein   \n",
       "\n",
       "           ensembl-id chromosome  \\\n",
       "0     ENSG00000183044         16   \n",
       "1     ENSG00000154263         17   \n",
       "2     ENSG00000179869          7   \n",
       "3     ENSG00000064687         19   \n",
       "4     ENSG00000164163          4   \n",
       "...               ...        ...   \n",
       "1157  ENSG00000196391         15   \n",
       "1158  ENSG00000170396          2   \n",
       "1159  ENSG00000151612          4   \n",
       "1160  ENSG00000130449          5   \n",
       "1161  ENSG00000174442         15   \n",
       "\n",
       "                                       genetic-category  gene-score  \\\n",
       "0        Rare Single Gene Mutation, Genetic Association         2.0   \n",
       "1                             Rare Single Gene Mutation         2.0   \n",
       "2                 Rare Single Gene Mutation, Functional         2.0   \n",
       "3                             Rare Single Gene Mutation         2.0   \n",
       "4                             Rare Single Gene Mutation         1.0   \n",
       "...                                                 ...         ...   \n",
       "1157                          Rare Single Gene Mutation         2.0   \n",
       "1158  Rare Single Gene Mutation, Genetic Association...         2.0   \n",
       "1159     Rare Single Gene Mutation, Genetic Association         2.0   \n",
       "1160  Rare Single Gene Mutation, Syndromic, Genetic ...         NaN   \n",
       "1161                          Rare Single Gene Mutation         2.0   \n",
       "\n",
       "      syndromic  eagle  number-of-reports  \n",
       "0             0    NaN                  7  \n",
       "1             0    NaN                  3  \n",
       "2             0    NaN                 11  \n",
       "3             0    NaN                  6  \n",
       "4             0    NaN                  3  \n",
       "...         ...    ...                ...  \n",
       "1157          0    NaN                  4  \n",
       "1158          0    NaN                 16  \n",
       "1159          0    NaN                  3  \n",
       "1160          1    NaN                  6  \n",
       "1161          0    NaN                  5  \n",
       "\n",
       "[1162 rows x 10 columns]"
      ]
     },
     "execution_count": 261,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sfari=pd.read_csv('SFARI-Gene_genes_01-16-2024release_01-19-2024export.csv')\n",
    "sfari"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>Probability_Class_0</th>\n",
       "      <th>Probability_Class_1</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>gene-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UBR5</td>\n",
       "      <td>ENSG00000104517</td>\n",
       "      <td>0.025221</td>\n",
       "      <td>0.974779</td>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHD9</td>\n",
       "      <td>ENSG00000177200</td>\n",
       "      <td>0.029630</td>\n",
       "      <td>0.970370</td>\n",
       "      <td>16</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHD6</td>\n",
       "      <td>ENSG00000124177</td>\n",
       "      <td>0.042962</td>\n",
       "      <td>0.957038</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>ENSG00000171634</td>\n",
       "      <td>0.043160</td>\n",
       "      <td>0.956840</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCOR2</td>\n",
       "      <td>ENSG00000196498</td>\n",
       "      <td>0.043535</td>\n",
       "      <td>0.956465</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18471</th>\n",
       "      <td>ADAM20</td>\n",
       "      <td>ENSG00000134007</td>\n",
       "      <td>0.992088</td>\n",
       "      <td>0.007912</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18472</th>\n",
       "      <td>MSMP</td>\n",
       "      <td>ENSG00000215183</td>\n",
       "      <td>0.992300</td>\n",
       "      <td>0.007700</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18473</th>\n",
       "      <td>ADAM30</td>\n",
       "      <td>ENSG00000134249</td>\n",
       "      <td>0.992557</td>\n",
       "      <td>0.007443</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18474</th>\n",
       "      <td>ADAM21</td>\n",
       "      <td>ENSG00000139985</td>\n",
       "      <td>0.992964</td>\n",
       "      <td>0.007036</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18475</th>\n",
       "      <td>GBP5</td>\n",
       "      <td>ENSG00000154451</td>\n",
       "      <td>0.993178</td>\n",
       "      <td>0.006822</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18476 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                1  Probability_Class_0  Probability_Class_1  \\\n",
       "0        UBR5  ENSG00000104517             0.025221             0.974779   \n",
       "1        CHD9  ENSG00000177200             0.029630             0.970370   \n",
       "2        CHD6  ENSG00000124177             0.042962             0.957038   \n",
       "3        BPTF  ENSG00000171634             0.043160             0.956840   \n",
       "4       NCOR2  ENSG00000196498             0.043535             0.956465   \n",
       "...       ...              ...                  ...                  ...   \n",
       "18471  ADAM20  ENSG00000134007             0.992088             0.007912   \n",
       "18472    MSMP  ENSG00000215183             0.992300             0.007700   \n",
       "18473  ADAM30  ENSG00000134249             0.992557             0.007443   \n",
       "18474  ADAM21  ENSG00000139985             0.992964             0.007036   \n",
       "18475    GBP5  ENSG00000154451             0.993178             0.006822   \n",
       "\n",
       "      Chromosome/scaffold name  gene-score  \n",
       "0                            8         2.0  \n",
       "1                           16         3.0  \n",
       "2                           20         NaN  \n",
       "3                           17         NaN  \n",
       "4                           12         NaN  \n",
       "...                        ...         ...  \n",
       "18471                       14         NaN  \n",
       "18472                        9         NaN  \n",
       "18473                        1         NaN  \n",
       "18474                       14         NaN  \n",
       "18475                        1         NaN  \n",
       "\n",
       "[18476 rows x 6 columns]"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results=pd.merge(results,sfari, left_on='1', right_on='ensembl-id',how='left')\n",
    "results=results[['0','1','Probability_Class_0','Probability_Class_1','Chromosome/scaffold name','gene-score']]\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0\n",
       "DUSP13B      2\n",
       "IL3RA        2\n",
       "CD99         2\n",
       "CRLF2        2\n",
       "CSF2RA       2\n",
       "            ..\n",
       "DAZ3         1\n",
       "IQGAP1       1\n",
       "NPIPB5       1\n",
       "KRTAP10-4    1\n",
       "GBP5         1\n",
       "Name: count, Length: 17881, dtype: int64"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['0'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Decile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "results['decile'] = pd.qcut(results['Probability_Class_1'], q=40, labels=['D{}'.format(i) for i in range(1, 41)])\n",
    "D1=results[results['decile']=='D1']\n",
    "D2=results[results['decile']=='D2']\n",
    "D40=results[results['decile']=='D40']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>Probability_Class_0</th>\n",
       "      <th>Probability_Class_1</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>gene-score</th>\n",
       "      <th>decile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>RAB40B</td>\n",
       "      <td>ENSG00000141542</td>\n",
       "      <td>0.213487</td>\n",
       "      <td>0.786513</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>RREB1</td>\n",
       "      <td>ENSG00000124782</td>\n",
       "      <td>0.213491</td>\n",
       "      <td>0.786509</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>RAB11B</td>\n",
       "      <td>ENSG00000185236</td>\n",
       "      <td>0.213517</td>\n",
       "      <td>0.786483</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>NCKAP5</td>\n",
       "      <td>ENSG00000176771</td>\n",
       "      <td>0.213826</td>\n",
       "      <td>0.786174</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>D40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0                1  Probability_Class_0  Probability_Class_1  \\\n",
       "458  RAB40B  ENSG00000141542             0.213487             0.786513   \n",
       "459   RREB1  ENSG00000124782             0.213491             0.786509   \n",
       "460  RAB11B  ENSG00000185236             0.213517             0.786483   \n",
       "461  NCKAP5  ENSG00000176771             0.213826             0.786174   \n",
       "\n",
       "    Chromosome/scaffold name  gene-score decile  \n",
       "458                       17         NaN    D40  \n",
       "459                        6         NaN    D40  \n",
       "460                       19         NaN    D40  \n",
       "461                        2         2.0    D40  "
      ]
     },
     "execution_count": 289,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "D40.tail(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>Probability_Class_0</th>\n",
       "      <th>Probability_Class_1</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>gene-score</th>\n",
       "      <th>decile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4619</th>\n",
       "      <td>PLOD3</td>\n",
       "      <td>ENSG00000106397</td>\n",
       "      <td>0.790703</td>\n",
       "      <td>0.209297</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4620</th>\n",
       "      <td>KLHDC3</td>\n",
       "      <td>ENSG00000124702</td>\n",
       "      <td>0.790781</td>\n",
       "      <td>0.209219</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4621</th>\n",
       "      <td>EIF3J</td>\n",
       "      <td>ENSG00000104131</td>\n",
       "      <td>0.790848</td>\n",
       "      <td>0.209152</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4622</th>\n",
       "      <td>DHX37</td>\n",
       "      <td>ENSG00000150990</td>\n",
       "      <td>0.790877</td>\n",
       "      <td>0.209123</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0                1  Probability_Class_0  Probability_Class_1  \\\n",
       "4619   PLOD3  ENSG00000106397             0.790703             0.209297   \n",
       "4620  KLHDC3  ENSG00000124702             0.790781             0.209219   \n",
       "4621   EIF3J  ENSG00000104131             0.790848             0.209152   \n",
       "4622   DHX37  ENSG00000150990             0.790877             0.209123   \n",
       "\n",
       "     Chromosome/scaffold name  gene-score decile  \n",
       "4619                        7         NaN    D30  \n",
       "4620                        6         NaN    D30  \n",
       "4621                       15         NaN    D30  \n",
       "4622                       12         NaN    D30  "
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "D30=results[results['decile']=='D30']\n",
    "D30.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export to csv\n",
    "results.to_csv('SVC_no_dup.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [],
   "source": [
    "D40.to_csv('D40.csv', index=False)\n",
    "D30.to_csv('D30.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
