{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Validation</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>PLATFORM</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>TYPESEQ</th>\n",
       "      <th>EFFECT</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>OMIM_ID</th>\n",
       "      <th>OMIM_PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-0006-003</td>\n",
       "      <td>2</td>\n",
       "      <td>98429018</td>\n",
       "      <td>98429019</td>\n",
       "      <td>CG</td>\n",
       "      <td>de novo</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CG</td>\n",
       "      <td>snp</td>\n",
       "      <td>splicing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM131</td>\n",
       "      <td>23505</td>\n",
       "      <td>615659.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-0028-003</td>\n",
       "      <td>10</td>\n",
       "      <td>28903528</td>\n",
       "      <td>28903533</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>de novo</td>\n",
       "      <td>TCACA</td>\n",
       "      <td>TCA</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>del</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>WAC</td>\n",
       "      <td>51322</td>\n",
       "      <td>615049.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1-0059-003</td>\n",
       "      <td>19</td>\n",
       "      <td>55686639</td>\n",
       "      <td>55686642</td>\n",
       "      <td>CG</td>\n",
       "      <td>inherited</td>\n",
       "      <td>CGA</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>del</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>SYT5</td>\n",
       "      <td>6861</td>\n",
       "      <td>600782.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1-0065-004</td>\n",
       "      <td>7</td>\n",
       "      <td>23191720</td>\n",
       "      <td>23191721</td>\n",
       "      <td>CG</td>\n",
       "      <td>de novo</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CG</td>\n",
       "      <td>snp</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>KLHL7</td>\n",
       "      <td>55975</td>\n",
       "      <td>611119.0</td>\n",
       "      <td>Retinitis pigmentosa 42, 612943 (3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1-0065-005</td>\n",
       "      <td>22</td>\n",
       "      <td>51019860</td>\n",
       "      <td>51019865</td>\n",
       "      <td>CG</td>\n",
       "      <td>inherited</td>\n",
       "      <td>CCAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>del</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>CHKB</td>\n",
       "      <td>1120</td>\n",
       "      <td>612395.0</td>\n",
       "      <td>Muscular dystrophy, congenital, megaconial typ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>AU4372309</td>\n",
       "      <td>11</td>\n",
       "      <td>94192626</td>\n",
       "      <td>94192627</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>de novo</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>snp</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>MRE11A</td>\n",
       "      <td>4361</td>\n",
       "      <td>600814.0</td>\n",
       "      <td>Ataxia-telangiectasia-like disorder, 604391 (3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>AU4376301</td>\n",
       "      <td>18</td>\n",
       "      <td>51795940</td>\n",
       "      <td>51795941</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>de novo</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>snp</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>POLI</td>\n",
       "      <td>11201</td>\n",
       "      <td>605252.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>AU4392301</td>\n",
       "      <td>11</td>\n",
       "      <td>46897495</td>\n",
       "      <td>46897496</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>de novo</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>snp</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>LRP4</td>\n",
       "      <td>4038</td>\n",
       "      <td>604270.0</td>\n",
       "      <td>?Myasthenic syndrome, congenital, 17, 616304 (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>AU4412302</td>\n",
       "      <td>17</td>\n",
       "      <td>27244344</td>\n",
       "      <td>27244351</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>de novo</td>\n",
       "      <td>AGGGGGG</td>\n",
       "      <td>AGGGGGGG</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>ins</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>PHF12</td>\n",
       "      <td>57649</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>AU4463303</td>\n",
       "      <td>6</td>\n",
       "      <td>42978705</td>\n",
       "      <td>42978710</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>de novo (mosaic?)</td>\n",
       "      <td>AAGAG</td>\n",
       "      <td>AAG</td>\n",
       "      <td>ILMN</td>\n",
       "      <td>del</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>PPP2R5D</td>\n",
       "      <td>5528</td>\n",
       "      <td>601646.0</td>\n",
       "      <td>Mental retardation, autosomal dominant 35, 616...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>226 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SAMPLE CHROM     START       END Platform         Validation  \\\n",
       "0    1-0006-003     2  98429018  98429019       CG            de novo   \n",
       "1    1-0028-003    10  28903528  28903533     ILMN            de novo   \n",
       "2    1-0059-003    19  55686639  55686642       CG          inherited   \n",
       "3    1-0065-004     7  23191720  23191721       CG            de novo   \n",
       "4    1-0065-005    22  51019860  51019865       CG          inherited   \n",
       "..          ...   ...       ...       ...      ...                ...   \n",
       "221   AU4372309    11  94192626  94192627     ILMN            de novo   \n",
       "222   AU4376301    18  51795940  51795941     ILMN            de novo   \n",
       "223   AU4392301    11  46897495  46897496     ILMN            de novo   \n",
       "224   AU4412302    17  27244344  27244351     ILMN            de novo   \n",
       "225   AU4463303     6  42978705  42978710     ILMN  de novo (mosaic?)   \n",
       "\n",
       "         REF       ALT PLATFORM TYPE   TYPESEQ                EFFECT  \\\n",
       "0          C         T       CG  snp  splicing                   NaN   \n",
       "1      TCACA       TCA     ILMN  del    exonic   frameshift deletion   \n",
       "2        CGA         C       CG  del    exonic   frameshift deletion   \n",
       "3          C         T       CG  snp    exonic              stopgain   \n",
       "4      CCAAA         C       CG  del    exonic   frameshift deletion   \n",
       "..       ...       ...      ...  ...       ...                   ...   \n",
       "221        G         A     ILMN  snp    exonic              stopgain   \n",
       "222        G         T     ILMN  snp    exonic              stopgain   \n",
       "223        C         T     ILMN  snp    exonic              stopgain   \n",
       "224  AGGGGGG  AGGGGGGG     ILMN  ins    exonic  frameshift insertion   \n",
       "225    AAGAG       AAG     ILMN  del    exonic   frameshift deletion   \n",
       "\n",
       "    GENE_SYMBOL  GENE_ID   OMIM_ID  \\\n",
       "0       TMEM131    23505  615659.0   \n",
       "1           WAC    51322  615049.0   \n",
       "2          SYT5     6861  600782.0   \n",
       "3         KLHL7    55975  611119.0   \n",
       "4          CHKB     1120  612395.0   \n",
       "..          ...      ...       ...   \n",
       "221      MRE11A     4361  600814.0   \n",
       "222        POLI    11201  605252.0   \n",
       "223        LRP4     4038  604270.0   \n",
       "224       PHF12    57649       NaN   \n",
       "225     PPP2R5D     5528  601646.0   \n",
       "\n",
       "                                        OMIM_PHENOTYPE  \n",
       "0                                                  NaN  \n",
       "1                                                  NaN  \n",
       "2                                                  NaN  \n",
       "3                  Retinitis pigmentosa 42, 612943 (3)  \n",
       "4    Muscular dystrophy, congenital, megaconial typ...  \n",
       "..                                                 ...  \n",
       "221    Ataxia-telangiectasia-like disorder, 604391 (3)  \n",
       "222                                                NaN  \n",
       "223  ?Myasthenic syndrome, congenital, 17, 616304 (...  \n",
       "224                                                NaN  \n",
       "225  Mental retardation, autosomal dominant 35, 616...  \n",
       "\n",
       "[226 rows x 16 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "duda=pd.read_excel('Duda_MSSNG.xlsx',sheet_name=\"Table S4\", skiprows=1)\n",
    "duda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n"
     ]
    }
   ],
   "source": [
    "mssng_sig=[\"CHD8\",\"ARID1B\",\"SCN2A\",\"DYRK1A\",\"SYNGAP1\",\"ADNP\",\"ANK2\",\"DSCAM\",\"ASH1L\",\"KMT5B\",\"KMT2A\",\n",
    "        \"CHD2\",\"GIGYF1\",\"GRIN2B\",\"POGZ\",\"ASXL3\",\"KMT2C\",\"WAC\",\"WDFY3\",\"SHANK3\",\"ANKRD11\",\"CACNA2D3\",\n",
    "        \"CUL3\",\"FOXP1\",\"KDM6B\",\"PHF2\",\"SPAST\",\"TBR1\",\"TCF7L2\",\"TNRC6B\",\"CIC\",\"CNOT3\",\"DDX3X\",\"DIP2C\",\n",
    "        \"MED13\",\"MYT1L\",\"NAA15\",\"PAX5\",\"PHF3\",\"PTEN\",\"SHANK2\",\"SMARCC2\",\"SRSF11\",\"UBN2\",\"UPF3B\",\"KDM6A\",\n",
    "        \"MECP2\",\"PCDH11X\",\"NEXMIF\",\"AFF2\",\"NLGN4X\",\"NLGN3\",\"FAM47A\",\"DYNC1H1\",\"SLC6A1\",\"AGAP2\",\n",
    "        \"ADCY3\",\"CLASP1\",\"DNMT3A\",\"MYO5A\",\"TAF6\"]\n",
    "\n",
    "print(len(mssng_sig))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n"
     ]
    }
   ],
   "source": [
    "novel=['CIC', 'CNOT3', 'DIP2C', 'MED13', 'PAX5', 'PHF3', 'SMARCC2', 'SRSF11', 'UBN2', 'DYNC1H1',\n",
    "        'AGAP2', 'ADCY3', 'CLASP1', 'MYO5A', 'TAF6', 'PCDH11X', 'NEXMIF','FAM47A']\n",
    "print(len(novel))  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "l=[]\n",
    "ENSG00000079432 --- CIC\n",
    "ENSG00000088038 --- CNOT3\n",
    "ENSG00000151240 --- DIP2C\n",
    "ENSG00000108510 --- MED13\n",
    "ENSG00000196092 --- PAX5\n",
    "ENSG00000118482 --- PHF3\n",
    "ENSG00000139613 --- SMARCC2\n",
    "ENSG00000116754 --- SRSF11\n",
    "ENSG00000157741 --- UBN2\n",
    "ENSG00000197102 --- DYNC1H1\n",
    "ENSG00000135439 --- AGAP2\n",
    "ENSG00000138031 --- ADCY3\n",
    "ENSG00000074054 --- CLASP1\n",
    "ENSG00000197535 --- MYO5A\n",
    "ENSG00000106290 --- \n",
    "ENSG00000099715 --- TAF6\n",
    "ENSG00000102290 ---\tPCDH11X\n",
    "ENSG00000050030 --- NEXMIF\n",
    "ENSG00000185448 --- FAM47A\n",
    "len(set(l))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Important\n",
    "changed KIAA2022 to NEXMIF because it is a synonym and KIAA2022 was not detected with Gprofiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "int"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "lof=list(set(duda[\"GENE_ID\"].astype(int).tolist()))\n",
    "\n",
    "complete_list = duda[\"GENE_ID\"].tolist()\n",
    "\n",
    "type(lof[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gmt_creator(gene_lists, name,query_data_type='ENTREZGENE_ACC'):\n",
    "\n",
    "\n",
    "    from gprofiler import GProfiler\n",
    "    gmt_strigs = []\n",
    "    \n",
    "    for gene_list, name in zip(gene_lists, name):\n",
    "        if isinstance(gene_list[0], str):  # Check if the first element of the gene list is a string\n",
    "            query_data_type = 'SYMBOL'\n",
    "        else:\n",
    "            query_data_type = 'ENTREZGENE_ACC'\n",
    "            \n",
    "        gp = GProfiler(return_dataframe=True)\n",
    "        gp_convert_df =gp.convert(organism  = 'hsapiens',\n",
    "                                    query= gene_list,\n",
    "                                        target_namespace= 'ENSG',\n",
    "                                        numeric_namespace=query_data_type,)\n",
    "        gp_convert_df = gp_convert_df[['incoming', 'converted']]\n",
    "        gp_convert_df = gp_convert_df.loc[gp_convert_df['converted'] != 'ENSG00000106290'] # X and Y chromosome error, removed the Y chromosome gene, as done before\n",
    "        ensemble_list = gp_convert_df['converted'].tolist()\n",
    "        gene_string = '\\t'.join(ensemble_list)\n",
    "        gmt_line= f'{name}\\tMutation: {name}\\t' + gene_string\n",
    "        gmt_strigs.append(gmt_line)\n",
    "        print(gmt_line)\n",
    "\n",
    "    with open('mssng.gmt', 'w') as f:\n",
    "        for gmt_line in gmt_strigs:\n",
    "            f.write(gmt_line + \"\\n\")     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSSNG_NOVEL\tMutation: MSSNG_NOVEL\tENSG00000079432\tENSG00000088038\tENSG00000151240\tENSG00000108510\tENSG00000196092\tENSG00000118482\tENSG00000139613\tENSG00000116754\tENSG00000157741\tENSG00000197102\tENSG00000135439\tENSG00000138031\tENSG00000074054\tENSG00000197535\tENSG00000099715\tENSG00000102290\tENSG00000050030\tENSG00000185448\n",
      "MSSNG_SIG\tMutation: MSSNG_SIG\tENSG00000100888\tENSG00000049618\tENSG00000136531\tENSG00000157540\tENSG00000197283\tENSG00000101126\tENSG00000145362\tENSG00000171587\tENSG00000116539\tENSG00000110066\tENSG00000118058\tENSG00000173575\tENSG00000146830\tENSG00000273079\tENSG00000143442\tENSG00000141431\tENSG00000055609\tENSG00000095787\tENSG00000163625\tENSG00000251322\tENSG00000167522\tENSG00000157445\tENSG00000036257\tENSG00000114861\tENSG00000285708\tENSG00000132510\tENSG00000197724\tENSG00000021574\tENSG00000136535\tENSG00000148737\tENSG00000100354\tENSG00000079432\tENSG00000088038\tENSG00000215301\tENSG00000151240\tENSG00000108510\tENSG00000186487\tENSG00000164134\tENSG00000196092\tENSG00000118482\tENSG00000171862\tENSG00000162105\tENSG00000139613\tENSG00000116754\tENSG00000157741\tENSG00000125351\tENSG00000147050\tENSG00000169057\tENSG00000099715\tENSG00000102290\tENSG00000050030\tENSG00000155966\tENSG00000146938\tENSG00000196338\tENSG00000185448\tENSG00000197102\tENSG00000157103\tENSG00000135439\tENSG00000138031\tENSG00000074054\tENSG00000119772\tENSG00000197535\n",
      "MSSNG_LOF\tMutation: MSSNG_LOF\tENSG00000119121\tENSG00000134532\tENSG00000183323\tENSG00000117139\tENSG00000142405\tENSG00000136158\tENSG00000137073\tENSG00000105722\tENSG00000183762\tENSG00000150938\tENSG00000131969\tENSG00000166148\tENSG00000029363\tENSG00000242265\tENSG00000148795\tENSG00000160862\tENSG00000100567\tENSG00000107731\tENSG00000116539\tENSG00000175216\tENSG00000089327\tENSG00000174032\tENSG00000198830\tENSG00000247596\tENSG00000085644\tENSG00000196865\tENSG00000160741\tENSG00000144504\tENSG00000102781\tENSG00000221909\tENSG00000189241\tENSG00000023287\tENSG00000164309\tENSG00000080603\tENSG00000100288\tENSG00000188419\tENSG00000142235\tENSG00000157741\tENSG00000164796\tENSG00000171862\tENSG00000196367\tENSG00000100815\tENSG00000169057\tENSG00000082781\tNone\tENSG00000079432\tENSG00000137204\tENSG00000086288\tENSG00000130940\tENSG00000215301\tENSG00000169718\tENSG00000157510\tENSG00000095787\tENSG00000170860\tENSG00000124214\tENSG00000149930\tENSG00000197299\tENSG00000112039\tENSG00000166436\tENSG00000176273\tENSG00000162825\tENSG00000137857\tENSG00000011201\tENSG00000049618\tENSG00000171943\tENSG00000168944\tENSG00000196408\tENSG00000136854\tENSG00000077232\tENSG00000132326\tENSG00000122550\tENSG00000139132\tENSG00000182898\tENSG00000122965\tENSG00000144285\tENSG00000138757\tENSG00000136531\tENSG00000134256\tENSG00000174238\tENSG00000027001\tENSG00000159069\tENSG00000114127\tENSG00000148773\tENSG00000075043\tENSG00000118058\tENSG00000213983\tENSG00000276293\tENSG00000129990\tENSG00000160310\tENSG00000133657\tENSG00000174946\tENSG00000155754\tENSG00000183530\tENSG00000128191\tENSG00000106804\tENSG00000262664\tENSG00000185800\tENSG00000164597\tENSG00000204060\tENSG00000147050\tENSG00000115919\tENSG00000163382\tENSG00000088038\tENSG00000108510\tENSG00000198885\tENSG00000138160\tENSG00000143373\tENSG00000140650\tENSG00000139211\tENSG00000149308\tENSG00000127080\tENSG00000140199\tENSG00000088812\tENSG00000092820\tENSG00000020922\tENSG00000119321\tENSG00000087303\tENSG00000119335\tENSG00000049089\tENSG00000172199\tENSG00000164134\tENSG00000100207\tENSG00000164830\tENSG00000181458\tENSG00000040633\tENSG00000188266\tENSG00000041982\tENSG00000145087\tENSG00000128595\tENSG00000109118\tENSG00000205726\tENSG00000173064\tENSG00000165671\tENSG00000053747\tENSG00000160058\tENSG00000115657\tENSG00000087470\tENSG00000184860\tENSG00000163214\tENSG00000107281\tENSG00000100888\tENSG00000171189\tENSG00000164151\tENSG00000171316\tENSG00000095002\tENSG00000266028\tENSG00000119285\tENSG00000164330\tENSG00000197748\tENSG00000239388\tENSG00000111863\tENSG00000101126\tENSG00000186364\tENSG00000174231\tENSG00000106049\tENSG00000011347\tENSG00000198752\tENSG00000251322\tENSG00000089169\tENSG00000180638\tENSG00000261787\tENSG00000148606\tENSG00000166845\tENSG00000157103\tENSG00000050344\tENSG00000115685\tENSG00000081014\tENSG00000082258\tENSG00000117620\tENSG00000005469\tENSG00000133794\tENSG00000112640\tENSG00000185024\tENSG00000162105\tENSG00000101331\tENSG00000048740\tENSG00000145907\tENSG00000162891\tENSG00000148120\tENSG00000118482\tENSG00000126524\tENSG00000141431\tENSG00000149532\tENSG00000198677\tENSG00000054392\tENSG00000197863\tENSG00000164342\tENSG00000197586\tENSG00000166164\tENSG00000130589\tENSG00000171798\tENSG00000084710\tENSG00000101751\tENSG00000060339\tENSG00000134569\tENSG00000139613\tENSG00000067900\tENSG00000175029\tENSG00000075568\tENSG00000179981\tENSG00000182698\tENSG00000163625\tENSG00000166348\tENSG00000198870\tENSG00000132849\tENSG00000050748\tENSG00000165487\tENSG00000065526\tENSG00000176994\tENSG00000188687\tENSG00000135387\tENSG00000121764\tENSG00000272886\tENSG00000205208\n"
     ]
    }
   ],
   "source": [
    "gmt_creator([novel, mssng_sig, lof], ['MSSNG_NOVEL', 'MSSNG_SIG','MSSNG_LOF'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test Gprofiler\n",
    " \n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "          \n",
    "   \n",
    "    \n",
    "     \n",
    "      \n",
    "       \n",
    "        \n",
    "         \n",
    "          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results=pd.read_csv('svc_clean.csv')\n",
    "gene_list=results['Ensembl_ID'].tolist()\n",
    "len(gene_list)\n",
    "\n",
    "\n",
    "data = pd.read_csv('svc_clean.csv')# chage data as need be\n",
    "\n",
    "data[\"Decile_rank\"] = pd.qcut(data[\"Probability_Class_1\"], 10, labels=False)\n",
    "data[\"Decile_rank\"] = data[\"Decile_rank\"].astype('category')\n",
    "data[\"Decile_rank\"] = data[\"Decile_rank\"].cat.rename_categories([f'D{i}' for i in range(10,0,-1)])\n",
    "\n",
    "deciles = {}\n",
    "for i in range(1, 11):\n",
    "        deciles[f'd_{i}'] = data[data[\"Decile_rank\"] == f'D{i}'].copy()\n",
    "\n",
    "\n",
    "\n",
    "d_1=deciles['d_1']['Ensembl_ID'].tolist()\n",
    "len(d_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from gprofiler import GProfiler\n",
    "\n",
    "def get_token_form_response(response):\n",
    "    if response.status_code == 200:\n",
    "        token = response.json()['organism']\n",
    "    else:\n",
    "        try:\n",
    "            error_message = 'Error: {}'.format(response.json()['message'])\n",
    "        except:\n",
    "            error_message = 'Error, status code {}'.format(response.status_code)\n",
    "        raise AssertionError(error_message)\n",
    "    #print(\"Token:\", token)\n",
    "    return token\n",
    "\n",
    "def main_custom(deciles, gmt_file,mutation_name):\n",
    "\n",
    "\n",
    "    dataframes= []\n",
    "    for i,key in enumerate(deciles):\n",
    "        decile_list = deciles[key][\"Ensembl_ID\"].to_list()\n",
    "        print(f'Processing {key,mutation_name}...')\n",
    "\n",
    "        gp = GProfiler(\n",
    "            user_agent='gprofiler_custom_gmt', #optional user agent\n",
    "            return_dataframe=True, #return pandas dataframe or plain python structures    \n",
    "        )\n",
    "\n",
    "        with open(gmt_file) as f:\n",
    "            response = requests.post('https://biit.cs.ut.ee/gprofiler/api/gost/custom/',\n",
    "                            json={'gmt':f.read(),\n",
    "                                'name': gmt_file})\n",
    "        token = get_token_form_response(response)\n",
    "\n",
    "        decile_results = gp.profile(query=decile_list, organism=token,user_threshold=1,domain_scope=\"custom\",background=d_1)\n",
    "        decile_results.insert(0, \"Decile\", i+1)\n",
    "        decile_results=decile_results[decile_results['native']==mutation_name]\n",
    "        dataframes.append(decile_results)\n",
    "\n",
    "\n",
    "    data = pd.concat(dataframes)\n",
    "    data= data.sort_values(by='Decile', ascending=False).reset_index(drop=True)\n",
    "    data['Intersection_to_Term_Ratio'] = data['intersection_size'] / data['term_size']\n",
    "    data=data.sort_values(by='Decile', ascending=True).reset_index(drop=True)\n",
    "\n",
    "    return data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing ('d_1', 'prb_rec_LGD')...\n",
      "Processing ('d_2', 'prb_rec_LGD')...\n",
      "Processing ('d_3', 'prb_rec_LGD')...\n",
      "Processing ('d_4', 'prb_rec_LGD')...\n",
      "Processing ('d_5', 'prb_rec_LGD')...\n",
      "Processing ('d_6', 'prb_rec_LGD')...\n",
      "Processing ('d_7', 'prb_rec_LGD')...\n",
      "Processing ('d_8', 'prb_rec_LGD')...\n",
      "Processing ('d_9', 'prb_rec_LGD')...\n",
      "Processing ('d_10', 'prb_rec_LGD')...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\joaop\\AppData\\Local\\Temp\\ipykernel_13580\\1555620758.py:43: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  data = pd.concat(dataframes)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Decile</th>\n",
       "      <th>source</th>\n",
       "      <th>native</th>\n",
       "      <th>name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>significant</th>\n",
       "      <th>description</th>\n",
       "      <th>term_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>query</th>\n",
       "      <th>parents</th>\n",
       "      <th>Intersection_to_Term_Ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>krish_duda</td>\n",
       "      <td>prb_rec_LGD</td>\n",
       "      <td>Mutation: prb_rec_LGD</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>Mutation: prb_rec_LGD</td>\n",
       "      <td>20</td>\n",
       "      <td>291</td>\n",
       "      <td>20</td>\n",
       "      <td>291</td>\n",
       "      <td>0.068729</td>\n",
       "      <td>1.0</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[]</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Decile      source       native                   name  p_value  \\\n",
       "0       1  krish_duda  prb_rec_LGD  Mutation: prb_rec_LGD      1.0   \n",
       "\n",
       "  significant            description term_size query_size intersection_size  \\\n",
       "0       False  Mutation: prb_rec_LGD        20        291                20   \n",
       "\n",
       "  effective_domain_size  precision  recall    query parents  \\\n",
       "0                   291   0.068729     1.0  query_1      []   \n",
       "\n",
       "  Intersection_to_Term_Ratio  \n",
       "0                        1.0  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MSSNG_lof = main_custom(deciles, 'krish_duda.gmt','prb_rec_LGD'),\"MSSNG_lof\" # native is the first \n",
    "MSSNG_lof[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
